pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -803,6 +803,330 @@ class StateVectorCudaMPI
+            data_type = CUDA_C_64F;
+        } else {
+            data_type = CUDA_C_32F;
+        }
+
+        std::vector<int> wires_int(wires.size());
+
+        // Transform indices between PL & cuQuantum ordering
+        std::transform(
+            wires.begin(), wires.end(), wires_int.begin(), [&](std::size_t x) {
+                return static_cast<int>(this->getTotalNumQubits() - 1 - x);
+            });
+
+        // split wires_int to global and local ones
+        std::vector<int> wires_local;
+        std::vector<int> wires_global;
+
+        for (const auto &wire : wires_int) {
+            if (wire < static_cast<int>(this->getNumLocalQubits())) {
+                wires_local.push_back(wire);
+            } else {
+                wires_global.push_back(wire);
+            }
+        }
+
+        std::vector<double> local_probabilities(Util::exp2(wires_local.size()));
+
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecAbs2SumArray(
+            /* custatevecHandle_t */ handle_.get(),
+            /* const void* */ BaseType::getData(),
+            /* cudaDataType_t */ data_type,
+            /* const uint32_t */ this->getNumLocalQubits(),
+            /* double* */ local_probabilities.data(),
+            /* const int32_t* */ wires_local.data(),
+            /* const uint32_t */ wires_local.size(),
+            /* const int32_t* */ maskBitString,
+            /* const int32_t* */ maskOrdering,
+            /* const uint32_t */ maskLen));
+
+        // create new MPI communicator groups
+        size_t subCommGroupId = 0;
+        for (size_t i = 0; i < wires_global.size(); i++) {
+            size_t mask = 1 << (wires_global[i] - this->getNumLocalQubits());
+            size_t bitValue = mpi_manager_.getRank() & mask;
+            subCommGroupId += bitValue
+                              << (wires_global[i] - this->getNumLocalQubits());
+        }
+        auto sub_mpi_manager0 =
+            mpi_manager_.split(subCommGroupId, mpi_manager_.getRank());
+
+        if (sub_mpi_manager0.getSize() == 1) {
+            return local_probabilities;
+        } else {
+            if (sub_mpi_manager0.getRank() == 0) {
+                subgroup_probabilities.resize(Util::exp2(wires_local.size()));
+            }
+
+            sub_mpi_manager0.Reduce<double>(local_probabilities,
+                                            subgroup_probabilities, 0, "sum");
+
+            return subgroup_probabilities;
+        }
+    }
+
+    /**
+     * @brief Utility method for samples.
+     *
+     * @param num_samples Number of Samples
+     *
+     * @return std::vector<size_t> A 1-d array storing the samples.
+     * Each sample has a length equal to the number of qubits. Each sample can
+     * be accessed using the stride sample_id*num_qubits, where sample_id is a
+     * number between 0 and num_samples-1.
+     */
+    auto generate_samples(size_t num_samples) -> std::vector<size_t> {
+        double epsilon = 1e-15;
```

Comment:
Do we need an eps value here? If so, can be programmatically pull in the macheps for the given type?

Reply:
yes, an eps is required here to ensure the success of custatevec API call. As required by the custatevecSamplerGetSquaredNorm API, it should be double.